### PR TITLE
ASoC: SOF: fix the potential memory leak

### PR DESCRIPTION
**Pull request by keqiaozhang**
Return statement causes loop exit, need to free dynamic allocated
buf before it.

Signed-off-by: Zhang Keqiao <keqiao.zhang@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -1844,7 +1844,8 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 			dev_err(sdev->dev, "error: unknown kcontrol type %d in widget %s\n",
 				widget->dobj.widget.kcontrol_type,
 				widget->name);
-			return -EINVAL;
+			ret = -EINVAL;
+			goto err;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: is this a real issue or a false-positive reported by a tool?

In this function, we loop twice on the same control components, and deal with the same switch case, so if there is an unknown kcontrol_type it'd be caught during the first iteration *before* allocating the memory, and we would never hit the second case.

Care to provide more details?

Comment (keqiaozhang) [reply sound/soc/sof/topology.c]: Thanks @plbossart 
This issue is reported by a tool.
You're right. Logically, it's impossible for this case to happen. I will mark it as 'not a problem' in the tool and close this PR.

Comment (keyonjie) [reply sound/soc/sof/topology.c]: > is this a real issue or a false-positive reported by a tool?
> 
> In this function, we loop twice on the same control components, and deal with the same switch case, so if there is an unknown kcontrol_type it'd be caught during the first iteration _before_ allocating the memory, and we would never hit the second case.

This is good point, we won't hit the default case in the 2nd iteration, maybe we should remove this part from the 2nd iteration:
```
-		default:
-			dev_err(sdev->dev, "error: unknown kcontrol type %d in widget %s\n",
-				widget->dobj.widget.kcontrol_type,
-				widget->name);
-			return -EINVAL;
```

And this will also fix the tool warning also I think.


> 
> Care to provide more details?

Comment (bardliao) [reply sound/soc/sof/topology.c]: Agree with @keyonjie , but may need a comment to explain why we don't need the default case at the 2nd iteration.

Comment (keqiaozhang) [reply sound/soc/sof/topology.c]: Thanks @keyonjie  and @bardliao 
Good suggestion, let's just remove the default case.

Comment (plbossart) [reply sound/soc/sof/topology.c]: removing the default case is going to trigger other warnings, it's better to mark this as not an error, or add the suggested fix with a note that this is not possible but added to make tools happy.

**Comment by keqiaozhang** [comment]
> This is a bit hard to follow, the PR title talks about fixing a memory leaks, where is that?
> 
> Plus, you mention case of unknown type is prevented by "blocked in the previous loop", but this is the first loop in the function. Which previous loop you refer to?

The PR title is changed to "ASoC: SOF: remove the redundant default case ...." above. 
Anyway, I will follow Pierre's suggestion to make things simple. Let's mark it as "not a problem" in the tool.

**Comment by kv2019i** [comment]
> The PR title is changed to "ASoC: SOF: remove the redundant default case ...." above.
> Anyway, I will follow Pierre's suggestion to make things simple. Let's mark it as "not a problem" in the tool.

@keqiaozhang  Ack, thanks!
